kasa/protocol.py

```diff
@@ -126,3 +134,152 @@ def decrypt(ciphertext: bytes) -> str:
         plaintext = bytes(buffer)
 
         return plaintext.decode()
+
+
+class TPLinkKLAP:
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, which appeared with firmware 1.1.0.
+    """
+
+    def __init__(self, host: str, authentication) -> None:
+        self.host = host
+        self.jar = aiohttp.CookieJar(unsafe=True, quote_cookie=False)
+        self.clientBytes = secrets.token_bytes(16)
+        self.authenticator = self.__computeAuthenticator(authentication)
+        self.handshake_lock = asyncio.Lock()
+        self.handshake_done = False
+
+        _LOGGER.debug("[KLAP] Created KLAP object for %s", self.host)
+
+    @staticmethod
+    def __computeAuthenticator(authentication: Optional[dict]) -> bytes:
+        if authentication is not None:
+            username = authentication["user"].encode()
+            password = authentication["password"].encode()
+        else:
+            username = bytearray(b"")
+            password = bytearray(b"")
+        _LOGGER.debug("[KLAP] Using username %s and password %s", username, password)
+        return hashlib.md5(
+            hashlib.md5(username).digest() + hashlib.md5(password).digest()
+        ).digest()
+
+    async def __handshake(self, session) -> None:
+        _LOGGER.debug("[KLAP] Starting handshake with %s", self.host)
+
+        # Handshake 1 has a payload of clientBytes
+        # and a response of 16 bytes, followed by sha256(clientBytes | authenticator)
+
+        url = "http://%s/app/handshake1" % self.host
+        resp = await session.post(url, data=self.clientBytes)
+        _LOGGER.debug("Got response of %d to handshake1", resp.status)
+        if resp.status != 200:
+            raise SmartDeviceException(
+                "Device responded with %d to handshake1" % resp.status
+            )
```

Comment:
```suggestion
            )

```
I'd separate the exit point here with an empty line.